lib/navigation/preprocess_encoders.cpp

```diff
@@ -33,10 +49,86 @@ core::EncoderData EncodersPreprocessor::detectOutliers(const core::EncoderData e
   -also has to be able to handle 1 unreliable sensor
   -also figure out return type/ what we update and update
   documentation as appropriate
+  }
   */
-  return {0, 0, 0, 0};
+template<typename T>
+std::array<core::Float,3> EncodersPreprocessor::quartiles(T encoder_data){
+  core::Float median;
+  core::Float q1;
+  core::Float q3;
+  core::Float iqr;
+  core::Float upper_bound;
+  core::Float lower_bound;
+  std::array<core::Float,3> output;
+  int q1_high = static_cast<int> (std::ceil((encoder_data.size()+1)/4.0));
+  int q1_low = static_cast<int> (std::floor((encoder_data.size()+1)/4.0));
+  int q3_high = static_cast<int> (std::ceil((3*(encoder_data.size()+1))/4.0));
+  int q3_low = static_cast<int> (std::ceil((3*(encoder_data.size()+1))/4.0));
+  q1 = encoder_data.at(q1_low-1) + (((encoder_data.size()+1)/4.0) - q1_low)*(encoder_data.at(q1_high-1) - encoder_data.at(q1_low-1));
+  q3 = encoder_data.at(q3_high-1) +(((3*(encoder_data.size()+1))/4.0) - q3_low)*(encoder_data.at(q3_high-1) - encoder_data.at(q3_low-1));
+  if(encoder_data.size() % 2 == 0){
+    median = (encoder_data.at((encoder_data.size()/2)-1) + encoder_data.at(((encoder_data.size()/2) + 1))-1)/2.0;
+  }
+  else{
+   median = encoder_data.at(((encoder_data.size()+1)/2)-1);
+  }
+  iqr = q3-q1;
+  upper_bound = median + 1.5*(iqr);
+  lower_bound = median - 1.5*(iqr);
+  output.at(0) = median;
+  output.at(1) = upper_bound;
+  output.at(2) = lower_bound;
+
+  return output; // return an output of the sequence median,upper_bound,lower_bound
+}
+
+
+core::EncoderData EncodersPreprocessor::detectOutliers(const core::EncoderData encoder_data){
+  const uint8_t num_reliable_encoders = std::accumulate(are_encoders_reliable_.begin(),are_encoders_reliable_.end(),0);
+  core::Float median;
```

Comment:
I feel like you've defined lots of medians in lots of different functions. The median should maybe be calculated once and then used to construct upper and lower bounds, nothing more.